pm/recipient_test.go

```diff
@@ -0,0 +1,851 @@
+	v.SetIsValidTicket(true)
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Test valid non-winning ticket
+	newSenderNonce := uint64(3)
+	ticket := newTicket(sender, params, newSenderNonce)
+
+	won, err := r.ReceiveTicket(ticket, sig, params.Seed)
+	if err != nil {
+		t.Error(err)
+	}
+	if won {
+		t.Errorf("expected valid non-winning ticket")
+	}
+
+	recipientRand := genRecipientRand(sender, secret, params.Seed)
+	senderNonce := r.(*recipient).senderNonces[recipientRand.String()]
+
+	if senderNonce != newSenderNonce {
+		t.Errorf("expected senderNonce to be %d, got %d", newSenderNonce, senderNonce)
+	}
+}
+
+func TestReceiveTicket_ValidWinningTicket(t *testing.T) {
+	sender, b, v, ts, faceValue, winProb, sig := newRecipientFixture(t)
+	secret := [32]byte{3}
+
+	r := NewRecipientWithSecret(b, v, ts, secret, faceValue, winProb)
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Test valid winning ticket
+	newSenderNonce := uint64(3)
+	ticket := newTicket(sender, params, newSenderNonce)
+
+	won, err := r.ReceiveTicket(ticket, sig, params.Seed)
+	if err != nil {
+		t.Error(err)
+	}
+	if !won {
+		t.Errorf("expected valid winning ticket")
+	}
+
+	recipientRand := genRecipientRand(sender, secret, params.Seed)
+	senderNonce := r.(*recipient).senderNonces[recipientRand.String()]
+
+	if senderNonce != newSenderNonce {
+		t.Errorf("expected senderNonce to be %d, got %d", newSenderNonce, senderNonce)
+	}
+
+	storeTickets, storeSigs, storeRecipientRands, err := ts.Load(ticket.RecipientRandHash.Hex())
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if len(storeTickets) != 1 {
+		t.Errorf("expected 1 stored tickets, got %d", len(storeTickets))
+	}
+	if len(storeSigs) != 1 {
+		t.Errorf("expected 1 stored sigs, got %d", len(storeSigs))
+	}
+	if len(storeRecipientRands) != 1 {
+		t.Errorf("expected 1 stored recipientRands, got %d", len(storeRecipientRands))
+	}
+
+	if storeTickets[0].Hash() != ticket.Hash() {
+		t.Errorf("expected store ticket hash %v, got %v", ticket.Hash(), storeTickets[0].Hash())
+	}
+
+	if !bytes.Equal(storeSigs[0], sig) {
+		t.Errorf("expected store sig 0x%x, got 0x%x", sig, storeSigs[0])
+	}
+
+	if crypto.Keccak256Hash(ethcommon.LeftPadBytes(storeRecipientRands[0].Bytes(), uint256Size)) != ticket.RecipientRandHash {
+		t.Error("expected store recipientRand to match ticket recipientRandHash")
+	}
+}
+
+func TestReceiveTicket_ValidWinningTicket_StoreError(t *testing.T) {
+	sender, b, v, ts, faceValue, winProb, sig := newRecipientFixture(t)
+	secret := [32]byte{3}
+
+	r := NewRecipientWithSecret(b, v, ts, secret, faceValue, winProb)
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub ticket store to fail store
+	ts.storeShouldFail = true
```

Comment:
IMO if you move this configuration to the line before calling the "code under test", the test becomes more readable, because this line is really the most different setup line compared to previous tests.

Reply:
To clarify: by moving this before the code under test do you mean `r.ReceiveTicket` in this case?

Reply:
Yes :)